# Develop Metric Data Analyst Business Case
In this notebook, I'm cleaning the xlsx data sets provided in order to facilitate their integration and analysis in Tableau

## Data Gathering
Let's read the different Excel sheets and load them into data frames

In [386]:
import pandas as pd

Case_Data_Analyst_xlsx = pd.ExcelFile("Tableau_Business_Case_Data_Analyst_July_2024_orig.xlsx")

### Dataset

In [387]:
dataset = pd.read_excel(Case_Data_Analyst_xlsx, sheet_name="Dataset")
dataset.head()

Project Name  \
0  Transforming a digital landscape: 7 years of l...   
1  Transforming a digital landscape: 7 years of l...   
2  Transforming a digital landscape: 7 years of l...   
3  Transforming a digital landscape: 7 years of l...   
4  Transforming a digital landscape: 7 years of l...   

                                             Excerpt  \
0  mSTAR/Bangladesh introduced the IFIC Amar Acco...   
1  mSTAR/Bangladesh organized a series of worksho...   
2  The process of digitizing transactions in USAI...   
3  mSTAR/Liberia played a key role in bringing to...   
4  mSTAR accompanied the staff from the Civil Ser...   

                                   All Interventions  Business Growth   \
0  Digital Microfinance (Loans or Credits), Finan...               NaN   
1  FinTech, Financial Inclusion, Digital Informat...               NaN   
2           Digital Finance, Mobile Banking Services               NaN   
3         Co-creation, Mobile Money, Digital Finance               NaN   
4     Co-creation, Mobile Money, Financial Inclusion               NaN   

   Improved Access to Digital Finance   Productivity Increased   \
0                                  4.0                      NaN   
1                                  NaN                      NaN   
2                                  4.0                      NaN   
3                                  4.0                      NaN   
4                                  NaN                      NaN   

   Improved Effectiveness and Efficiency  \
0                                    NaN   
1                                    NaN   
2                                    4.0   
3                                    NaN   
4                                    NaN   

   Improved Governance and Institutional Capacity  
0                                             NaN  
1                                             NaN  
2                                             NaN  
3                                             NaN  
4                                             NaN

In [388]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 8 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   Project Name                                    303 non-null    object 
 1   Excerpt                                         303 non-null    object 
 2   All Interventions                               303 non-null    object 
 3   Business Growth                                 13 non-null     float64
 4   Improved Access to Digital Finance              122 non-null    float64
 5   Productivity Increased                          7 non-null      float64
 6   Improved Effectiveness and Efficiency           21 non-null     float64
 7   Improved Governance and Institutional Capacity  23 non-null     float64
dtypes: float64(5), object(3)
memory usage: 19.1+ KB


### Taxonomy

In [389]:
taxonomy = pd.read_excel(Case_Data_Analyst_xlsx, sheet_name="Taxonomy")
taxonomy

First Order Code  \
0                                    Interventions   
1                                  Digital Finance   
2                                              NaN   
3                                              NaN   
4                                              NaN   
5                                              NaN   
6                                              NaN   
7                                              NaN   
8                                              NaN   
9                                              NaN   
10                                             NaN   
11                                        Outcomes   
12                                 Business Growth   
13              Improved Access to Digital Finance   
14                          Productivity Increased   
15           Improved Effectiveness and Efficiency   
16  Improved Governance and Institutional Capacity   

                          Second Order Code         Third Order Code  
0                                       NaN                      NaN  
1                                       NaN                      NaN  
2                       Financial Inclusion                      NaN  
3                                   FinTech                      NaN  
4                                       NaN             Mobile Money  
5                                       NaN  Mobile Banking Services  
6   Digital Microfinance (Loans or Credits)                      NaN  
7                                       NaN                 Payments  
8                         Digital Insurance                      NaN  
9                               Investments                      NaN  
10     Digital Financial Market Regulations                      NaN  
11                                      NaN                      NaN  
12                                      NaN                      NaN  
13                                      NaN                      NaN  
14                                      NaN                      NaN  
15                                      NaN                      NaN  
16                                      NaN                      NaN

### Project Metadata

In [390]:
project_metadata = pd.read_excel(Case_Data_Analyst_xlsx, sheet_name="Project Metadata")
project_metadata.head()

Project Names  \
0  Transforming a digital landscape: 7 years of l...   
1  Local development II - provincial (LD II - P) ...   
2  Communication for healthy living : final repor...   
3  Local development II urban project : final rep...   
4  Final performance evaluation of the Haiti inte...   

                                Country  Project Start Date  Project End Date  
0  Burma | India | Liberia | Bangladesh                2012              2019  
1                                 Egypt                1988              1992  
2                                 Egypt                2003              2010  
3                                 Egypt                1989              1992  
4                                 Haiti                2009              2015

## Data Cleaning
Some of the data sets requires some cleaning to facilitate their integration and analysis in Tableau
### Dataset

In [391]:
dataset_cleaned = dataset.copy()

#### Split and explode "All Interventions"

In [392]:
# Split "All Interventions" around ", " and save the list in "Intervention"
dataset_cleaned["Intervention"] = dataset_cleaned["All Interventions"].str.split(",")

# Unpivot/explode the "Intervention" cells into multiple rows (one per value in the split list)
dataset_cleaned = dataset_cleaned.explode("Intervention")

dataset_cleaned.reset_index(drop=True, inplace=True)

In [393]:
dataset_cleaned.head()

Project Name  \
0  Transforming a digital landscape: 7 years of l...   
1  Transforming a digital landscape: 7 years of l...   
2  Transforming a digital landscape: 7 years of l...   
3  Transforming a digital landscape: 7 years of l...   
4  Transforming a digital landscape: 7 years of l...   

                                             Excerpt  \
0  mSTAR/Bangladesh introduced the IFIC Amar Acco...   
1  mSTAR/Bangladesh introduced the IFIC Amar Acco...   
2  mSTAR/Bangladesh introduced the IFIC Amar Acco...   
3  mSTAR/Bangladesh organized a series of worksho...   
4  mSTAR/Bangladesh organized a series of worksho...   

                                   All Interventions  Business Growth   \
0  Digital Microfinance (Loans or Credits), Finan...               NaN   
1  Digital Microfinance (Loans or Credits), Finan...               NaN   
2  Digital Microfinance (Loans or Credits), Finan...               NaN   
3  FinTech, Financial Inclusion, Digital Informat...               NaN   
4  FinTech, Financial Inclusion, Digital Informat...               NaN   

   Improved Access to Digital Finance   Productivity Increased   \
0                                  4.0                      NaN   
1                                  4.0                      NaN   
2                                  4.0                      NaN   
3                                  NaN                      NaN   
4                                  NaN                      NaN   

   Improved Effectiveness and Efficiency  \
0                                    NaN   
1                                    NaN   
2                                    NaN   
3                                    NaN   
4                                    NaN   

   Improved Governance and Institutional Capacity  \
0                                             NaN   
1                                             NaN   
2                                             NaN   
3                                             NaN   
4                                             NaN   

                              Intervention  
0  Digital Microfinance (Loans or Credits)  
1                      Financial Inclusion  
2                          Digital Finance  
3                                  FinTech  
4                      Financial Inclusion

In [394]:
dataset_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 765 entries, 0 to 764
Data columns (total 9 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   Project Name                                    765 non-null    object 
 1   Excerpt                                         765 non-null    object 
 2   All Interventions                               765 non-null    object 
 3   Business Growth                                 39 non-null     float64
 4   Improved Access to Digital Finance              313 non-null    float64
 5   Productivity Increased                          20 non-null     float64
 6   Improved Effectiveness and Efficiency           53 non-null     float64
 7   Improved Governance and Institutional Capacity  66 non-null     float64
 8   Intervention                                    765 non-null    object 
dtypes: float64(5), object(4)
memory usage: 53.9+

We can notice that the number of entries went from 303 to 765 by splitting "All Interventions" into multiple rows

#### Clean "Intervention"
It seems that the split left some leading white space in the Intervention.

In [395]:
dataset_cleaned["Intervention"].unique()

array(['Digital Microfinance (Loans or Credits)', ' Financial Inclusion',
       ' Digital Finance', 'FinTech', ' Digital Information Services',
       ' Network', ' Mobile Money', 'Digital Finance',
       ' Mobile Banking Services', 'Co-creation', ' Digital Platform',
       'Mobile Money', ' ', ' Payments', '', ' Digital Identity Systems',
       ' Co-creation', ' Cross-Border Data Flows', ' Access to ICTs',
       ' Digital Inclusion', 'Management Information Systems',
       'Automation', ' Digital Skills/Capacity Building',
       'Digital Financial Market Regulations', 'Payments',
       'Financial Inclusion', ' FinTech', 'Digital Platform',
       ' Gender Digital Divide', 'Mobile Banking Services',
       'Access to ICTs', ' Digital Economy', ' Data Systems',
       ' Age Digital Divide', ' Digital Literacy',
       ' Digital Microfinance (Loans or Credits)',
       ' Internet and Data Governance', ' Interoperability',
       ' Management Information Systems',
       ' Digital

In [396]:
dataset_cleaned[dataset_cleaned["Intervention"].str.startswith(" ")].shape

(458, 9)

In [397]:
dataset_cleaned["Intervention"] = dataset_cleaned["Intervention"].str.lstrip(" ")

In [398]:
dataset_cleaned["Intervention"].unique()

array(['Digital Microfinance (Loans or Credits)', 'Financial Inclusion',
       'Digital Finance', 'FinTech', 'Digital Information Services',
       'Network', 'Mobile Money', 'Mobile Banking Services',
       'Co-creation', 'Digital Platform', '', 'Payments',
       'Digital Identity Systems', 'Cross-Border Data Flows',
       'Access to ICTs', 'Digital Inclusion',
       'Management Information Systems', 'Automation',
       'Digital Skills/Capacity Building',
       'Digital Financial Market Regulations', 'Gender Digital Divide',
       'Digital Economy', 'Data Systems', 'Age Digital Divide',
       'Digital Literacy', 'Internet and Data Governance',
       'Interoperability', 'Digital Data Collection',
       'Management (e-administration)', 'Digital Maintenance',
       'E-procurement', 'Digital Insurance', 'Connectivity',
       'Loans and Credit', 'Financial Technology', 'e-Service Delivery',
       'Engagement (e-participation)', 'Digital Security',
       'Research and Develop

In [399]:
assert dataset_cleaned[dataset_cleaned["Intervention"].str.startswith(" ")].shape[0] == 0

In [400]:
dataset_cleaned.head()

Project Name  \
0  Transforming a digital landscape: 7 years of l...   
1  Transforming a digital landscape: 7 years of l...   
2  Transforming a digital landscape: 7 years of l...   
3  Transforming a digital landscape: 7 years of l...   
4  Transforming a digital landscape: 7 years of l...   

                                             Excerpt  \
0  mSTAR/Bangladesh introduced the IFIC Amar Acco...   
1  mSTAR/Bangladesh introduced the IFIC Amar Acco...   
2  mSTAR/Bangladesh introduced the IFIC Amar Acco...   
3  mSTAR/Bangladesh organized a series of worksho...   
4  mSTAR/Bangladesh organized a series of worksho...   

                                   All Interventions  Business Growth   \
0  Digital Microfinance (Loans or Credits), Finan...               NaN   
1  Digital Microfinance (Loans or Credits), Finan...               NaN   
2  Digital Microfinance (Loans or Credits), Finan...               NaN   
3  FinTech, Financial Inclusion, Digital Informat...               NaN   
4  FinTech, Financial Inclusion, Digital Informat...               NaN   

   Improved Access to Digital Finance   Productivity Increased   \
0                                  4.0                      NaN   
1                                  4.0                      NaN   
2                                  4.0                      NaN   
3                                  NaN                      NaN   
4                                  NaN                      NaN   

   Improved Effectiveness and Efficiency  \
0                                    NaN   
1                                    NaN   
2                                    NaN   
3                                    NaN   
4                                    NaN   

   Improved Governance and Institutional Capacity  \
0                                             NaN   
1                                             NaN   
2                                             NaN   
3                                             NaN   
4                                             NaN   

                              Intervention  
0  Digital Microfinance (Loans or Credits)  
1                      Financial Inclusion  
2                          Digital Finance  
3                                  FinTech  
4                      Financial Inclusion

### Taxonomy
In order to use the hierarchical structure of "Intervention", I need to extract them from the data set and fill the null values
#### Extract Intervention
We'll extract the entries between the "First Order Code" entry "Interventions" and "Outcomes" excluded since all the other entries do not interest us

In [401]:
taxonomy_cleaned = taxonomy.copy()
intervention_idx = taxonomy_cleaned[taxonomy_cleaned["First Order Code"] == "Interventions"].index[0]
print(intervention_idx)

0


In [402]:
outcomes_idx = taxonomy_cleaned[taxonomy_cleaned["First Order Code"] == "Outcomes"].index[0]
print(outcomes_idx)

11


In [403]:
taxonomy_cleaned = taxonomy_cleaned.iloc[(intervention_idx + 1): outcomes_idx]
taxonomy_cleaned

First Order Code                        Second Order Code  \
1   Digital Finance                                      NaN   
2               NaN                      Financial Inclusion   
3               NaN                                  FinTech   
4               NaN                                      NaN   
5               NaN                                      NaN   
6               NaN  Digital Microfinance (Loans or Credits)   
7               NaN                                      NaN   
8               NaN                        Digital Insurance   
9               NaN                              Investments   
10              NaN     Digital Financial Market Regulations   

           Third Order Code  
1                       NaN  
2                       NaN  
3                       NaN  
4              Mobile Money  
5   Mobile Banking Services  
6                       NaN  
7                  Payments  
8                       NaN  
9                       NaN  
10                      NaN

#### Filling NaN
In order to use the hierarchical structure in Tableau, I decided to fill all the missing order codes
##### Propagate code to the right
Let's first propagate the existing First Order Code the the Second and Third Order

In [404]:
first_order_exists_mask = taxonomy_cleaned["First Order Code"].notna()
print(first_order_exists_mask)
taxonomy_cleaned.loc[first_order_exists_mask, ["Second Order Code", "Third Order Code"]] = taxonomy_cleaned.loc[first_order_exists_mask, "First Order Code"]
taxonomy_cleaned

1      True
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
Name: First Order Code, dtype: bool


First Order Code                        Second Order Code  \
1   Digital Finance                          Digital Finance   
2               NaN                      Financial Inclusion   
3               NaN                                  FinTech   
4               NaN                                      NaN   
5               NaN                                      NaN   
6               NaN  Digital Microfinance (Loans or Credits)   
7               NaN                                      NaN   
8               NaN                        Digital Insurance   
9               NaN                              Investments   
10              NaN     Digital Financial Market Regulations   

           Third Order Code  
1           Digital Finance  
2                       NaN  
3                       NaN  
4              Mobile Money  
5   Mobile Banking Services  
6                       NaN  
7                  Payments  
8                       NaN  
9                       NaN  
10                      NaN

Let's do the same for the Second order toward the Third order

In [405]:
second_order_exists_mask = taxonomy_cleaned["Second Order Code"].notna()
print(second_order_exists_mask)
taxonomy_cleaned.loc[second_order_exists_mask, "Third Order Code"] = taxonomy_cleaned.loc[second_order_exists_mask, "Second Order Code"]
taxonomy_cleaned

1      True
2      True
3      True
4     False
5     False
6      True
7     False
8      True
9      True
10     True
Name: Second Order Code, dtype: bool


First Order Code                        Second Order Code  \
1   Digital Finance                          Digital Finance   
2               NaN                      Financial Inclusion   
3               NaN                                  FinTech   
4               NaN                                      NaN   
5               NaN                                      NaN   
6               NaN  Digital Microfinance (Loans or Credits)   
7               NaN                                      NaN   
8               NaN                        Digital Insurance   
9               NaN                              Investments   
10              NaN     Digital Financial Market Regulations   

                           Third Order Code  
1                           Digital Finance  
2                       Financial Inclusion  
3                                   FinTech  
4                              Mobile Money  
5                   Mobile Banking Services  
6   Digital Microfinance (Loans or Credits)  
7                                  Payments  
8                         Digital Insurance  
9                               Investments  
10     Digital Financial Market Regulations

##### Fill all NaN with previous values in each column

In [406]:
taxonomy_cleaned.fillna(method='ffill', inplace=True)

/var/folders/hr/2w5fdz6n7n7cznb6zr5sklvw0000gn/T/ipykernel_55165/515369521.py:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  taxonomy_cleaned.fillna(method='ffill', inplace=True)


## Save cleaned data
Let's save the 2 cleaned Dataset and Taxonomy data frames and the untouched Project Metadata to a new Excel file

In [408]:
# dataset_cleaned.to_excel("Tableau_Business_Case_Data_Analyst_July_2024_dataset_cleaned.xlsx", sheet_name="Dataset", index=False)
with pd.ExcelWriter("Tableau_Business_Case_Data_Analyst_July_2024_cleaned.xlsx") as excel_writer:
    dataset_cleaned.to_excel(excel_writer, sheet_name="Dataset", index=False)
    taxonomy_cleaned.to_excel(excel_writer, sheet_name="Taxonomy", index=False)
    project_metadata.to_excel(excel_writer, sheet_name="Project Metadata", index=False)